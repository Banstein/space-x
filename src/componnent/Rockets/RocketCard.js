import React from 'react';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { reserveRocket, cancelRocketBooking } from '../../redux/rockets/rockets';

const RocketCard = ({
  id, description, name, image, reserved,
}) => {
  const dispatch = useDispatch();
  const reservingRocket = (e) => {
    const { id } = e.target;
    dispatch(reserveRocket(+id));
  };

  const cancelReservation = (e) => {
    const { id } = e.target;
    dispatch(cancelRocketBooking(+id));
  };

  return (
    <div className="rocket-container flex-center">
      <div className="rocket-image">
        <img src={image} alt={name} />
      </div>
      <div className="rocket-info">
        <h3 className="rocket-name">{name}</h3>
        <p className="description">
          {reserved && <span className="reserved">Reserved</span>}
          {description}
        </p>
        {!reserved && (
          <button
            type="button"
            id={id}
            className="reserve-btn clickable"
            onClick={reservingRocket}
          >
            Reserve Rocket
          </button>
        )}
        {reserved && (
          <button
            type="button"
            id={id}
            className="cancel-btn clickable"
            onClick={cancelReservation}
          >
            Cancel Reservation
          </button>
        )}
      </div>
    </div>
  );
};

RocketCard.propTypes = {
  id: PropTypes.number.isRequired,
  description: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  reserved: PropTypes.bool.isRequired,
};

export default RocketCard;
