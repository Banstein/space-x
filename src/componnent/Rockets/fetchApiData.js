import axios from 'axios';

const url = 'https://api.spacexdata.com/v3/rockets';

const fetchRocketsApiData = async () => {
  const getData = await axios.get(url);
  const Rockets = [];

  getData.data.forEach((rocketItem) => {
    const flickrImage = rocketItem.flickr_images[0];
    Rockets.push({
      id: rocketItem.id,
      rocket_name: rocketItem.rocket_name,
      description: rocketItem.description,
      image: flickrImage,
      reserved: false,
    });
  });

  return Rockets;
};

export default fetchRocketsApiData;
