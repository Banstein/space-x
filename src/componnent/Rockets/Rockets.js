import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { rocketDispatcher } from '../../redux/rockets/rockets';
import RocketCard from './RocketCard';

const Rockets = () => {
  const dispatch = useDispatch();
  const rocketsData = useSelector((state) => state.rocketsReducer);

  useEffect(() => {
    if (!rocketsData.isDataStored) {
      dispatch(rocketDispatcher());
    }
  }, []);

  const rockets = rocketsData.data;
  let rocketsList = [];
  if (rockets) {
    rocketsList = rockets.map((rocket) => (
      <li key={rocket.id} className="rocket-item">
        <RocketCard
          className="rocket-card"
          description={rocket.description}
          image={rocket.image}
          name={rocket.rocket_name}
          id={rocket.id}
          reserved={rocket.reserved}
        />
      </li>
    ));
  }

  return (
    <section>
      <div className="rockets">
        <ul classeName="rocket-list">
          {rocketsList}
        </ul>
      </div>
    </section>
  );
};

export default Rockets;
