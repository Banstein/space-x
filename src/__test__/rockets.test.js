import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import store from '../redux/configureStore';
import Rockets from '../componnent/Rockets/Rockets';
import RocketCard from '../componnent/Rockets/RocketCard';

it('Test if Rocket render correctly', () => {
  const rocket = {
    id: 1,
    description: 'This is the Falcon1 Rocket',
    image: 'https://imgur.com/DaCfMsj.jpg',
    name: 'Rocket',
    reserved: false,
  };

  const tree = renderer.create(
    <Provider store={store}>
      <RocketCard
        description={rocket.description}
        image={rocket.image}
        name={rocket.name}
        id={rocket.id}
        reserved={rocket.reserved}
      />
    </Provider>,
  );
  expect(tree).toMatchSnapshot();
});

it('Test if Rockets renders correctly', () => {
  const tree = renderer.create(
    <Provider store={store}>
      <Rockets />
    </Provider>,
  );
  expect(tree).toMatchSnapshot();
});
