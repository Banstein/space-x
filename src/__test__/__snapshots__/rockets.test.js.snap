// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test if Rocket render correctly 1`] = `
<div
  className="rocket-container flex-center"
>
  <div
    className="rocket-image"
  >
    <img
      alt="Rocket"
      src="https://imgur.com/DaCfMsj.jpg"
    />
  </div>
  <div
    className="rocket-info"
  >
    <h3
      className="rocket-name"
    >
      Rocket
    </h3>
    <p
      className="description"
    >
      This is the Falcon1 Rocket
    </p>
    <button
      className="reserve-btn clickable"
      id={1}
      onClick={[Function]}
      type="button"
    >
      Reserve Rocket
    </button>
  </div>
</div>
`;

exports[`Test if Rockets renders correctly 1`] = `
<section>
  <div
    className="rockets"
  >
    <ul
      className="rocket-list"
    />
  </div>
</section>
`;